includes/class-sensei-course.php

```diff
@@ -3068,40 +3069,40 @@ public static function is_prerequisite_complete( $course_id ){
     }// end is_prerequisite_complete
 
 	/**
-	 * Allowing user to set course archive page as front page
+	 * Allowing user to set course archive page as front page.
+	 *
+	 * Expects to be called during pre_get_posts, but only if page_on_front is set
+	 * to a non-empty value.
 	 *
 	 * @since 1.9.5
 	 * @param WP_Query $query hooked in from pre_get_posts
 	 */
 	function allow_course_archive_on_front_page( $query ) {
+		// Flag used to indicate if our test query is running: we use this instead of temporarily
+		// hooking/unhooking due to a core WP bug @see https://core.trac.wordpress.org/ticket/17817
+		static $running = false;
 
-		global $wp;
-
-		if ( ! is_page() ){
-			return; // This only needs to run on the home page. See https://github.com/Automattic/sensei/issues/1438
+		// Bail if it's clear we're not looking at a static front page or if the $running flag is
+		// set @see https://github.com/Automattic/sensei/issues/1438
+		if ( $running || ! $query->is_main_query() || ! is_page() || is_admin() ) {
```

Comment:
So keeping `$running` in its current position seems sensible, since it doesn't entail a function call and is a pre-requisite to everything else. I also like the positioning of `! $query->is_main_query()` because for all but one query in any given request we'll short-circuit here.

It might make sense to remove `is_admin()` from this method altogether and instead unhook this callback completely when `admin_init` fires, though, thus sparing us these tests altogether during admin requests (or, we could use a different action - `init` or something else - as a stepping stone: ie if it's not a frontend request, don't hook up this action).

What are your thoughts on that? My own view is that could be done as part of a subsequent change distinct from this one (because the fact that this callback is being hooked up during admin requests is not new), but I'd love your feedback.